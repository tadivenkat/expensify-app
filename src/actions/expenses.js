import database from '../firebase/firebase';

export const addExpenseAction = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        payload: expense
    };
};

export const startAddExpenseAction = (expenseData = {}) => {
    const {description = '', amount = 0, note = '', createdAt = 0} = expenseData;
    const expense = {description, amount, note, createdAt};
    return (dispatch) => {
        return database.ref('expenses').push(expense).then((data) => {
            dispatch(addExpenseAction({
                id: data.key,
                ...expense
            }));
        }).catch((error) => {
            console.log("Error adding expense", error);
        });
    }
};

export const removeExpenseAction = ({id} = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id 
    }
};

export const editExpenseAction = ({id, amount, description, note, createdAt} = {}) => {
    return {
        type: 'EDIT_EXPENSE',
        payload: {
            id,
            amount,
            description,
            note,
            createdAt
        }
    }
};

export const setExpensesAction = (expenses = []) => {
    return {
        type: 'SET_EXPENSES',
        payload: expenses
    }
};

export const startSetExpensesAction = () => {
    const expenses = [];
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpensesAction(expenses));
        }).catch((error) => {
            console.log("Error retreiving expenses", error);
        });
    };
};