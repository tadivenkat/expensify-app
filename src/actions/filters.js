
export const setTextFilterAction = (text = '') => {
    return {
        type: 'SET_TEXT_FILTER',
        payload: text
    }
};

export const setSortByFilterAction = (sortBy = '') => {
    return {
        type: 'SET_SORT_BY',
        payload: sortBy
    };
}

export const setStartDateFilterAction = (startDate) => {
    return {
        type: 'SET_START_DATE',
        payload: startDate
    };
}

export const setEndDateFilterAction = (endDate) => {
    return {
        type: 'SET_END_DATE',
        payload: endDate
    }
}