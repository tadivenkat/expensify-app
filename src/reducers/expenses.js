
export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.payload];
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.payload.id) {
                    return {...expense, ...action.payload};
                } else {
                    return expense;
                }
            });
        default: 
            return state;
    }
};