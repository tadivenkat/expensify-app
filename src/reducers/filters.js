
const filtersDefaultState = {
    text: '',
    sortBy: 'date',
    startDate: undefined,
    endDate: undefined
}

export default (state = filtersDefaultState, action) => {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.payload
            };
        case 'SET_SORT_BY':
            return {
                ...state,
                sortBy: action.payload
            };
        case 'SET_START_DATE':
            return {
                ...state,
                startDate: action.payload
            };
        case 'SET_END_DATE':
            return {
                ...state,
                endDate: action.payload
            };
        default: 
            return state;
    }
};