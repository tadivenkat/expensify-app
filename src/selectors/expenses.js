// Return subset of expenses applying the filters.

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        let startDateMatch = false;
        let endDateMatch = false;
        const textMatch = expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;
        if (!!startDate) {
            if (!!expense.createdAt) {
                startDateMatch = startDate <= expense.createdAt;
            } else {
                startDateMatch = true;
            }
        } else {
            startDateMatch = true;
        }
        if (!!endDate) {
            if (!!expense.createdAt) {
                endDateMatch = endDate >= expense.createdAt;
            } else {
                endDateMatch = true;
            }
        } else {
            endDateMatch = true;
        }
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expense1, expense2) => {
        switch (sortBy) {
            case 'amount':
                return expense1.amount - expense2.amount;
            case 'date':
                return expense1.createdAt - expense2.createdAt;
            default:
                return expense1.amount - expense2.amount;
        }
    });
};