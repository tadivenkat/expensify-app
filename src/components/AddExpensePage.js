import React from 'react';
import {connect} from 'react-redux';
import {startAddExpenseAction} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';
import Header from './Header';

export class AddExpensePage extends React.Component {
        onSubmit = (expense) => {
        this.props.onSubmit(expense);
        this.props.history.push('/dashboard');            
    }    
    render() {
        return (
            <div>
                <Header/>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        onSubmit: (expense) => dispatch(startAddExpenseAction(expense))
    };
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);