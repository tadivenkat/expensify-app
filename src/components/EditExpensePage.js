import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpenseAction, removeExpenseAction} from '../actions/expenses';

const EditExpensePage = (props) => {
    return (
        <div>
            <h1>Editing Expense: {props.match.params.id}</h1>
            <ExpenseForm 
                expense={props.expense}
                onSubmit={(expense) => {
                    console.log(expense);
                    props.dispatch(editExpenseAction(expense));
                    props.history.push('/');
                }}
            />
            <button onClick={() => {
                props.dispatch(removeExpenseAction(props.expense));
                props.history.push("/");
            }}>Remove</button>
        </div>
    );
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

export default connect(mapStateToProps)(EditExpensePage);