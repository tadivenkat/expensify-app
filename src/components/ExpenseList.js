import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import filterExpenses from '../selectors/expenses';
import {totalExpenditure} from '../selectors/expenses';
import moment from 'moment';
import numeral from 'numeral';


export const ExpenseList = (props) => {
    const {text, sortBy, startDate, endDate} = props.filters;
    return (
        <div>
            <h1>Expense List</h1>
            <h4>Filters Applied: Text: {text}, SortBy: {sortBy}, startDate: {moment(startDate).format('DD-MMM-YYYY')}, endDate: {moment(endDate).format('DD-MMM-YYYY')}</h4>
            <h2>You have {props.expenses.length} {props.expenses.length > 1 ? 'expenses' : 'expense'} with a total expenditure of {numeral(totalExpenditure(props.expenses)/100).format('$ 0,0.00')}</h2>
            {props.expenses.map((expense) => <ExpenseListItem 
                                                    key={expense.id} 
                                                    expense={expense}/>)}
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        expenses: filterExpenses(state.expenses, state.filters),
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseList);