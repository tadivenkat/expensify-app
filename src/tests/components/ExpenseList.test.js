import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseList} from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';
import filters from '../fixtures/filters';

import toJSON from 'enzyme-to-json';

test('Testing ExpenseList Component Snapshot', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} filters={filters}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Testing ExpenseList Component Snapshot with empty expenses as input', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} filters={filters}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});