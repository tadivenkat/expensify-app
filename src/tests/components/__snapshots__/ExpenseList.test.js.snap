// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ExpenseList Component Snapshot 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <h4>
    Filters Applied: Text: 
    , SortBy: 
    amount
    , startDate: 
    01-Jan-1970
    , endDate: 
    01-Jan-1970
  </h4>
  <h2>
    You have 
    4
     expenses with a total expenditure of 
    $ 760.00
  </h2>
  <ExpenseListItem
    expense={
      Object {
        "amount": 10000,
        "createdAt": 8000,
        "description": "pen",
        "id": "1",
        "note": "test note for test expense",
      }
    }
    key="1"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 25000,
        "createdAt": -35,
        "description": "apple",
        "id": "2",
        "note": "test note for test expense",
      }
    }
    key="2"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 6000,
        "createdAt": 5000,
        "description": "orange",
        "id": "3",
        "note": "test note for test expense",
      }
    }
    key="3"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 35000,
        "createdAt": 200,
        "description": "mango",
        "id": "4",
        "note": "test note for test expense",
      }
    }
    key="4"
  />
</div>
`;

exports[`Testing ExpenseList Component Snapshot with empty expenses as input 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <h4>
    Filters Applied: Text: 
    , SortBy: 
    amount
    , startDate: 
    01-Jan-1970
    , endDate: 
    01-Jan-1970
  </h4>
  <h2>
    You have 
    0
     expenses with a total expenditure of 
    $ 0.00
  </h2>
</div>
`;
